File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.cc
#include <cstdio>

int arr[1024][1024];

int main() {
    int n, m;

    scanf("%d%d", &n, &m);

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            int temp;

            scanf("%d", &temp);

            arr[i][j] = temp;
        }
    }

    for (int i = 1; i < n; i++)
        for (int j = 0; j < n; j++)
            arr[i][j] += arr[i - 1][j];

    for (int i = 0; i < n; i++)
        for (int j = 1; j < n; j++)
            arr[i][j] += arr[i][j - 1];

    for (int i = 0; i < m; i++) {
        int x1, x2, y1, y2;
        int result;

        scanf("%d%d%d%d", &x1, &y1, &x2, &y2);

        result = arr[x2 - 1][y2 - 1];

        if ((x1 > 1) && (y1 > 1))
            result += arr[x1 - 2][y1 - 2];

        if (x1 > 1)
            result -= arr[x1 - 2][y2 - 1];

        if (y1 > 1)
            result -= arr[x2 - 1][y1 - 2];
        
        printf("%d\n", result);
    }

    return 0;
}

File: 백준/Silver/11660. 구간 합 구하기 5/구간 합 구하기 5.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int n, m;
	static int[][] arr;
	static StringBuilder sb = new StringBuilder();
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		arr = new int[n+1][n+1];

		for(int i=1; i<=n; i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=1; j<=n; j++) {
				arr[i][j] = arr[i][j-1] + Integer.parseInt(st.nextToken());
			}
		}
		
		for(int i=1; i<=n; i++) {
			for(int j=1; j<=n; j++) {
				arr[j][i] += arr[j-1][i];
			}
		}
		
		for(int i=0; i<m; i++) {
			st = new StringTokenizer(br.readLine());
			int ay = Integer.parseInt(st.nextToken());
			int ax = Integer.parseInt(st.nextToken());
			int by = Integer.parseInt(st.nextToken());
			int bx = Integer.parseInt(st.nextToken());
			
			int result = arr[by][bx] - arr[by][ax-1] - arr[ay-1][bx] + arr[ay-1][ax-1];
			sb.append(result).append("\n");
		}
		
		System.out.println(sb);
	}

}
